Add support for async checkpointing (#13658)
Lightning-AI/lightning: docs/source-pytorch/common/checkpointing_expert.rst

       respectively, common for most use cases.
   * - :class:`~pytorch_lightning.plugins.io.XLACheckpointIO`
     - CheckpointIO that utilizes :func:`xm.save` to save checkpoints for TPU training strategies.
   * - :class:`~pytorch_lightning.plugins.io.HPUCheckpointIO`
     - CheckpointIO to save checkpoints for HPU training strategies.
   * - :class:`~pytorch_lightning.plugins.io.AsyncCheckpointIO`

>      - ``AsyncCheckpointIO`` enables saving the checkpoints asynchronously using the ``ThreadPoolExecutor``.

Reviewer: ```suggestion
     - ``AsyncCheckpointIO`` enables saving the checkpoints asynchronously in a thread.
```
I would not mention the specific method we use. We are also only using a pool size of 1 as a special case.

.. note::

    Some ``TrainingTypePlugins`` like ``DeepSpeedStrategy`` do not support custom ``CheckpointIO`` as checkpointing logic is not modifiable.

**************************

> Asynchronous Checkpointing

Reviewer 2: Did you do any benchmarking?

Author: very basic one, just on BoringModel() trained on 1000 epochs. It was around 30% faster when I added some sleep time while saving the checkpoint. Do you have any recommendations?

**************************

.. warning::

    This is currently an experimental plugin/feature and API changes are to be expected.

To enable saving the checkpoints asynchronously without blocking your training, you can configure
:class:`~pytorch_lightning.plugins.io.async_plugin.AsyncCheckpointIO` plugin to ``Trainer``.

> It uses its base ``CheckpointIO`` plugin's saving logic to save the checkpoint but performs this operation asynchronously.

Reviewer: I would not mention any "base" checkpoint io or wrapping logic here, because it's not needed in the code sample below. Only mention it in the second paragraph where it matters.

.. note::

    Some ``TrainingTypePlugins`` like ``DeepSpeedStrategy`` do not support custom ``CheckpointIO`` as checkpointing logic is not modifiable.

**************************
Asynchronous Checkpointing
**************************

.. warning::

    This is currently an experimental plugin/feature and API changes are to be expected.

To enable saving the checkpoints asynchronously without blocking your training, you can configure
:class:`~pytorch_lightning.plugins.io.async_plugin.AsyncCheckpointIO` plugin to ``Trainer``.

> It uses it's base ``CheckpointIO`` plugin's saving logic to save the checkpoint but performs the this operation asynchronously.

Reviewer 3: ```suggestion
It uses its base ``CheckpointIO`` plugin's saving logic to save the checkpoint but performs this operation asynchronously.
```

By default, this base ``CheckpointIO`` will be set-up for you and all you need to provide is the ``AsyncCheckpointIO`` instance to the ``Trainer``.

.. code-block:: python

   from pytorch_lightning.plugins.io import AsyncCheckpointIO

   async_ckpt_io = AsyncCheckpointIO()
   trainer = Trainer(plugins=[async_ckpt_io])

>    model = LitModel()
>    trainer.fit(model)

Reviewer 4: nit: this doesn't seem relevant

```suggestion
```

> If you want to plugin your own base ``CheckpointIO`` and want it be behave asynchronously, pass it while initializing ``AsyncCheckpointIO``.

Reviewer 3: ```suggestion
If you want the plugin to use your own base ``CheckpointIO`` and want the base to behave asynchronously, pass it as an argument while initializing ``AsyncCheckpointIO``.
```
